certora/specs/AccessManager.spec

```
@@ -0,0 +1,588 @@
+
+    // values after
+    mathint scheduleAfter = getSchedule(e2, operationId);
+
+    // transition
+    assert scheduleBefore != scheduleAfter => (
+        scheduleBefore + expiration() >  e1.block.timestamp + 0 && // +0 cast to mathint
+        scheduleBefore + expiration() <= e2.block.timestamp + 0 && // +0 cast to mathint
+        scheduleAfter == 0
+    );
+}
+
+rule getScheduleChangeCall(bytes32 operationId) {
+    env e;
+
+    // values before
+    mathint scheduleBefore = getSchedule(e, operationId);
+
+    // arbitrary function call
+    method f; calldataarg args; f(e, args);
+
+    // values after
+    mathint scheduleAfter = getSchedule(e, operationId);
+
+    // transitions
+    assert scheduleBefore != scheduleAfter => (
+        (f.selector == sig:schedule(address,bytes,uint48).selector    && scheduleAfter >= e.block.timestamp + 0) ||
+        (f.selector == sig:execute(address,bytes).selector            && scheduleAfter == 0                    ) ||
+        (f.selector == sig:cancel(address,address,bytes).selector     && scheduleAfter == 0                    ) ||
+        (f.selector == sig:consumeScheduledOp(address,bytes).selector && scheduleAfter == 0                    ) ||
+        (isOnlyAuthorized(f)                                          && scheduleAfter == 0                    )
+    );
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Functions: restricted functions can only be called by owner                                                         │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule restrictedFunctions(env e) {
+    require nonpayable(e);
+    require envsanity(e);
+
+    method f;
+    calldataarg args;
+
+    f(e,args);
+
+    assert (
+        f.selector == sig:labelRole(uint64,string).selector ||
+        f.selector == sig:setRoleAdmin(uint64,uint64).selector ||
+        f.selector == sig:setRoleGuardian(uint64,uint64).selector ||
+        f.selector == sig:setGrantDelay(uint64,uint32).selector ||
+        f.selector == sig:setTargetAdminDelay(address,uint32).selector ||
+        f.selector == sig:updateAuthority(address,address).selector ||
+        f.selector == sig:setTargetClosed(address,bool).selector ||
+        f.selector == sig:setTargetFunctionRole(address,bytes4[],uint64).selector
+    ) => (
+        hasRole_1(e, ADMIN_ROLE(), e.msg.sender)
+    );
+}
+
+rule restrictedFunctionsGrantRole(env e) {
+    require nonpayable(e);
+    require envsanity(e);
+
+    uint64 roleId;
+    address account;
+    uint32 executionDelay;
+
+    // We want to check that the caller has the admin role before we possibly grant it.
+    bool hasAdminRoleBefore = hasRole_1(e, getRoleAdmin(roleId), e.msg.sender);
+
+    grantRole(e, roleId, account, executionDelay);
+
+    assert hasAdminRoleBefore;
```

I don't completely get how this rule for `grantRole` works (and the `revokeRole` one). If `hasAdminRoleBefore` changes, shouldn't we assert why it changed? For example:

```suggestion
    assert !hasAdminAfter => roleId == ADMIN_ROLE()
```

Probably a bit more complex than that but that'll be the idea

---

The purpose of this rule is only to assert that if `grantRole` successfully executes, the caller is an admin.

So for this rule we don't care about `hasAdminAfter`.